Add native histogram specification

beorn7 opened the pull request:
Here is finally the opus magnum that was in the making for far too long. Partially addressing https://github.com/prometheus/prometheus/issues/11561 .

It is partially more documentation than specification, but I assume the specification section is still the least bad place for it.

This is intentionally not yet complete. TODOs mark the missing parts, most of which cannot be resolved at the moment because implementation work is still missing and/or decisions still have to be made.

The intention here is to finally get something out here, even if it is pretty much still a "living document". It will probably take us another year or so to resolve all TODOs.

Dear reviewers, primary goal of the review is to detect what I got wrong here and what I have forgotten. Language improvements are of course welcome, but of secondary importance. And as said above, the TODOs should mostly stay for this first version. Of course, if you have an easy way of resolving one, be my guest.

Tagging potential reviewers here (simply focus on your areas of expertise):
@bwplotka @juliusv @krajorama @zenador @fatsheep9146 @NeerajGartia21 @Maniktherana @fpetkovski @vesari @carrieedwards @fionaliao @codesome @ArthurSens @csmarchbanks @fstab

beorn7 commented:
https://deploy-preview-2539--prometheus-docs.netlify.app/docs/specs/native_histograms/ is the preview, very convenient to see the rendered layout.

krajorama commented:
cc @jhesketh  @charleskorn have also dealt with native histograms due to their query engine work.

krajorama commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+
+Each span consists of a pair of numbers, a signed 32-bit integer (short: int32)
+called _offset_ and an unsigned 32-bit integer (short: uint32) called _length_.
+Only the first span in each list can have a negative offset. It defines the
```
For NHCB not even the first offset can be negative as the offset is interpreted as an offset into the custom values.

krajorama reviewed (commented):
Couple of comments. Haven't gone through yet even once, on line 738 Instrumentation libraries.

krajorama commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+As with classic histograms, an estimation of the maximum and minimum
+observation in a histogram can be performed using 1 and 0, respectively, as the
+first parameter of `histogram_quantile`. This estimation is much more useful
+with native histograms, though, not only because of the usually higher
```
native histograms with standard schemas

krajorama commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+
+`histogram_count()` and `histogram_sum()` return the count of observations or the sum of
+observations, respectively, that are contained in a native histogram. As they are normal
+function, their result cannot be used in a range selector. Instead of using
```
```suggestion
functions, their result cannot be used in a range selector. Instead of using
```

krajorama commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+## Prometheus UI
+
+This section describes the rendering of histograms by Prometheus's own UI. This
+MAY be used as a guideline for external graphing frontends (like Perses or
+Grafana).
```
```suggestion
MAY be used as a guideline for third party graphing frontends.
```

krajorama commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+TODO: Is there more to say? I'm not a remote write & read specialist. Does
+remote read really work?
+
+TODO: What are the plans to auto-convert classic histograms to NHCBs in
```
No plan at the moment because we cannot guarantee that the classic histogram came from scrape where we for sure know we have all the buckets present. This is true for both receiving and sending remote write. In fact one of the reason for having NHCB is to avoid scenarios where classic histograms are not consistent (missing/late due to independent series). Therefor currently the only plans are to have NHCB when we know we have all data : scrape - being implemented, receiving OTEL histogram - issue there, expose from instrumentation directly.

krajorama commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+In [remote write v2](https://prometheus.io/docs/specs/remote_write_spec_2_0/),
+native histograms are a stable feature.
+
+TODO: Is there more to say? I'm not a remote write & read specialist. Does
```
Remote read should work , either it sends chunks or converts the result: https://github.com/prometheus/prometheus/blob/fc3d19725d83e13363ce40d56672c6f97e2955e7/storage/remote/codec.go#L155

Remove write works for sure, it's in use.

The one TODO I think is to verify/fix the situation when exemplars of a native histogram are sent in two requests instead of one - we might consider the first batch out-of-order. But it has been a while since I worked on this so not 100% sure it's a problem.

krajorama commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+to match schema 8. (In the unlikely case that a schema smaller than -4 is used,
+the ingestion will fail.)
+
+TODO: Is the OTLP receiver documented anywhere? Link the documentation here.
```
+ TODO convert to NHCB explicit bucket histograms. https://github.com/prometheus/prometheus/issues/15022

krajorama replied:
@aknuds1  about documentation

krajorama commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+period, which comes at the cost of collecting and storing classic and native
+histograms in parallel for a while.
+
+The first step is to update the instrumentation to expose classic and native
```
or start scraping NHCB on top of classic histograms

krajorama commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+histogram query without any consideration about the right switch-over.
+
+Once there is confidence that all queries have been migrated correctly,
+configure Prometheus to only scrape native histograms (which is the “normal”
```
or start dropping some metrics with relabel rules to make this incremental

krajorama commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+setting). If everything still works, it is time to remove classic histograms
+from the instrumentation.
+
+The Grafana Mimir documentation contains [a detailed migration
```
thanks!

krajorama reviewed (commented):
done with first read through

fionaliao commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+With Prometheus v2.49 and later, it is possible to finetune the scrape protocol
+negotiation globally or per scrape config via the `scrape_protocols` config
+setting. It is a list defining the content negotiation priorities. Its default
+value depends on the `--enable-feature=native-histograms` flag. If the flage is
```
```suggestion
value depends on the `--enable-feature=native-histograms` flag. If the flag is
```

fionaliao commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+
+### Finetuning content negotiation
+
+With Prometheus v2.49 and later, it is possible to finetune the scrape protocol
```
```suggestion
With Prometheus v2.49 and later, it is possible to fine-tune the scrape protocol
```

fionaliao commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+negotiation globally or per scrape config via the `scrape_protocols` config
+setting. It is a list defining the content negotiation priorities. Its default
+value depends on the `--enable-feature=native-histograms` flag. If the flage is
+set, it is `[ PrometheusProto, OpenMetricsText1.0.0, OpenMetricsText0.0.1,
+PrometheusText0.0.4 ]`, otherwise `[ OpenMetricsText1.0.0,
+OpenMetricsText0.0.1, PrometheusText0.0.4 ]`. This results in the behavior
```
Should the diff between the lists be explicitly pointed out - i.e. point out the difference is `PrometheusProto`. Just so you don't need to jump between the lists to figure out the change.

fionaliao commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+  (because the leading byte has to denote the different type).
+- A counter reset for a counter histogram (to be stored in the leading byte as
+  counter reset information, see details below).
+- A schema change (which means we need a new chunk layout, and one and the same
+  chunk can only have one chunk layout).
```
(Opinion) I think this reads better:

```suggestion
- A schema change (which means we need a new chunk layout, as a 
  chunk can only have one chunk layout).
```

fionaliao commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+mark a histogram time series as stale. However, this would require cutting a
+new chunk, just for the purpose of marking the series as stale, because a float
+value following a histogram value has to be stored in a different chunk (see
+above). Therefore, there is also a histogram version of a stale marker where
+the field for the sum of observations is set to the special stale `NaN` value.
+In this case, all other fields are ignored, so they can be set to values
```
Having NaN observations could also result in the sum of observations being NaN according to
https://github.com/prometheus/docs/blob/2d3624ced474a0eeb70ec9a63d1d4ef7c96283a3/content/docs/specs/native_histograms.md?plain=1#L492-L496

How are the two cases (staleness markers vs NaN observations) distinguished? For the NaN observations case we wouldn't want to ignore the other fields.

krajorama replied:
I would guess count==0 is the distinguishing feature, so not all fields are ignored

bboreham replied:
There are lots of NaN values; we pick [one of them](https://github.com/prometheus/prometheus/blob/cd1f8ac129a289be8e7d98b6de57a9ba5814c406/model/value/value.go#L24-L28) to mean "stale".
It is not expected that the same value can be generated by accident.

beorn7 replied:
The count value shouldn't matter. If the sum is the specific bit pattern that defines a staleness marker, the whole histogram must be treated as a staleness marker.

krajorama replied:
This is super confusing :( Not sure how we can improve the text, but it seems like having `NaN` written here is the source of trouble. Let me suggest something in a new comment to try and clarify.

beorn7 replied:
I'll resolve this conversation, because we have a new one below with @krajorama 's suggestion.

fionaliao commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+resulting histogram. The removal is only precise if the threshold coincides
+with a bucket boundary. Otherwise, interpolation within the affected buckets
+has to be used, as described [above](#interpolation-within-a-bucket). The
+counter vs. gauge flavor of the histogram is preserved. (TODO: These operaters
```
```suggestion
counter vs. gauge flavor of the histogram is preserved. (TODO: These operators
```

carrieedwards commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+- The count of observations, encoded as varbit-uint for the 1st sample and as
+  varbit-int for any further samples, using the same “delta of deltas” approach
+  as for timestamps.
+- The zero bucket population, enccoded as varbit-uint for the 1st sample and as
```
```suggestion
- The zero bucket population, encoded as varbit-uint for the 1st sample and as
```

carrieedwards commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+Neighboring buckets are merged to reduce the schema, and regular buckets are
+merged with the zero bucket to increase the width of the zero bucket.
+
+At this point in the procedure, both histograms have the schema and zero bucket
```
```suggestion
At this point in the procedure, both histograms have the same schema and zero bucket
```

carrieedwards commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+TODO: Currently, Prometheus returns a `CounterResetHint` of
+`UnknownCounterReset` for the 1st histogram of a chunk marked as
+`NotCounterReset` (which is essentially situation (4) from above). Once [this
+issue](https://github.com/prometheus/prometheus/issues/15247) is resolves, we
```
```suggestion
issue](https://github.com/prometheus/prometheus/issues/15247) is resolved, we
```

carrieedwards commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+buckets of all histograms in the chunk, so cutting a new chunk enables a
+simpler set of buckets for the new chunk.
+
+This in turn implies that there will never be a counter reset within a chunk.
+The only histogram for which counter reset information has to be persisted is
+therefore the 1st histogram in a chunk. This happens in the so-called
```
I think this phrasing is a little bit confusing. Maybe something like this?

```suggestion
This in turn implies that there will never be a counter reset after the first sample in a chunk. Therefore, the only counter reset information that has to be persisted is is that of the first histogram in the chunk. This happens in the so-called
```

carrieedwards commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+
+An observation of `NaN` goes into no bucket, but increments the count of
+observations. This implies that the count of observations can be greater than
+the sum of all buckets, and the difference is the number of `NaN` observations.
```
Should it mention that for histograms that haven't had any NaN observations/have a sum of NaN, the count has to be equal to the sum of overall counts in the negative buckets, positive buckets, and zero count?

carrieedwards commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+OTel has a dense rather than a sparse representation of buckets. One might see
+OTel as “Prom with only one span”.
+
+The Prom _zero bucket_ is called _zero count_ for OTel, but works the same,
```
I believe the attribute is called ZeroCount in both the Histogram/Float Histogram struct in Prometheus and the OTel Exponential Histogram

beorn7 commented:
Thanks everyone for the review so far.

In the commit I have just pushed I have addressed all the feedback and resolved the conversation where I assumed that my response is uncontroversial. The conversations still open might require more discussion.

I have _not yet_ reworded the counter reset handling according to our recent discussions. I will do that tomorrow with a refreshed brain.

beorn7 commented:
OK, counter reset handling should be up to date now.

krajorama commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+snapshot (Go type `record.Type`): `record.HistogramSamples` (string
+representation `histogram_samples`, numerical value 7) for integer histograms,
+and `record.FloatHistogramSamples` (string representation
+`float_histogram_samples`, numerical value 8) for float histograms.
```
https://github.com/prometheus/prometheus/pull/15467

beorn7 replied:
Thank you. I added the new record types.

krajorama commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+changes could be part of a compatible resolution reduction (which happens
+regularly to [reduce the bucket count of a
+histogram](#limit-the-bucket-count)). Both of the following is true for a
+compatible resolution reduction:
```
on second read it's ok to have here, but I don't see it in the code

beorn7 replied:
It's kind of distributed in various parts of the appender (and the "ok to append" checks) in the tsdb code. I also believe that the current ingestion code doesn't cover all edge cases. I have noted down a TODO on my list to verify and fix that. (The `DetectReset` method MUST result in the exact same counter reset detection as the one we do on the fly during ingestion. The reason why we don't just call `DetectReset` on each ingested histogram is performance.)

beorn7 commented:
Note: Apparently, I forgot to send all the 52 pending replies I wrote several weeks ago.
They are submitted now, so now it makes way more sense where I resolved a conversation and where not.

beorn7 commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+exposition format. See details in the [documentation for the
+`native-histograms` feature
+flag](https://prometheus.io/docs/prometheus/latest/feature_flags/#native-histograms).
+(TODO: Normalization will become the default in v3. Update this as appropriate.)
```
I have inverted the wording now, so that everything is from the perspective of v3 being the reference point.

krajorama commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+current release v2.55 and v3.00. Update this section with the stable release,
+once it has happened.)
+
+Due to the pervasive nature of the native-histogram related changes, the
```
lgtm

krajorama commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+
+(TODO: Update section as progress is made.)
+
+## Instrumentation libraries
```
yes please, I think one sentence about not forgetting exemplars would be appreciated

beorn7 replied:
I even added two paragraphs now, also extending a bit about how instrumentation libraries can track classic and native histograms in parallel and that the API (`Observe(...)`) doesn't change between them.

krajorama commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+memory available to the program. Therefore, an instrumentation library SHOULD
+offer a bucket limitation strategy. It MAY set one by default, depending on the
+typical use cases the library is used for. (TODO: Maybe we should say that a
+strategy SHOULD be set by default. The Go library is currently not limiting the
```
ack, let's remember this

krajorama commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2385 @@
+new chunk, just for the purpose of marking the series as stale, because a float
+value following a histogram value has to be stored in a different chunk (see
+above). Therefore, there is also a histogram version of a stale marker where
+the field for the sum of observations is set to the special stale `NaN` value.
```
```suggestion
the field for the sum of observations is set to the special stale `NaN` value. Note: this special `NaN` value is a float value different to other representations of `NaN`, in particular other `NaN` values that would arise from arithmetic operations, thus it is different from observing a "natural" `NaN` value as described in [Special cases of observed values](#special-cases-of-observed-values).
```

beorn7 replied:
I have tried to explain this in the first sentence of this section (albeit in a more condensed form). The confusion encountered here proves that we need a better explanation. However, I think it should be at the beginning of the section, not in the middle. See what I have done (in the next commit).

krajorama replied:
thank you

beorn7 commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2353 @@
+be gauges. Adding and subtracting counter histograms requires caution, but
+PromQL allows it. Adding a gauge histogram and a counter histogram results in a
+gauge histogram. Adding two counter histograms with contradicting counter reset
+hints triggers a warn-level annotation. (TODO: The latter not yet implemented.
```
I assume the explanation above is satisfactory.

beorn7 commented:
OK, I think we are done. @krajorama please let me know if I have missed anything.

bwplotka commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2416 @@
+the compression of float samples, and Prometheus v3 might very well adopt this
+approach again in the future.)
+
+In a counter histogram, the total _count_ of observation and the counts in the
```
I would expect this section to tell us what to do for gauge histograms. Are there representable by native histograms? If yes, then how?

beorn7 replied:
The concept of a gauge histogram is not linked to NH. It was explained (and referenced) above. There, it is explicitly said that a NH can have gauge flavor. Not sure what else there is to explain for the representation. (Should we say somewhere that in a gauge histogram, all the counts are gauges? That's kind of the idea of a gauge histogram and not specific to native histograms, so maybe we shouldn't create the impression it is. Note that this whole paragraph here exists because classic histograms were treating _sum_ wrongly from the beginning, and it got finally fixed with NH.)

bwplotka commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2416 @@
+defines the way bucket boundaries are calculated. The currently valid values
+are -53 and the range between and including -4 and +8. More schemas may be
+added in the future. -53 is a schema for so-called _custom bucket boundaries_
+or short _custom buckets_, while the other schema numbers represent the
```
Let's add those names to glossary 🤗

beorn7 replied:
Maybe we can consider extending the glossary to all the new terms introduced by this document. However, there would be a lot. If this was a printed document, it might still be helpful for cross reference, but since this is electronic, the reader will simply search for any term they don't remember. (The few terms in the glossary so far are those that might easily be misunderstood or confused.)

bwplotka commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2416 @@
+- A (possibly empty) list of _custom values_.
+- _Exemplars_.
+
+### Flavors
```
Perhaps standard schema and non-standard schemas is another dimension worth calling out?

bwplotka replied:
fair!

bwplotka commented on content/docs/specs/native_histograms.md:
````diff
@@ -0,0 +1,2416 @@
+// [...]
+```
+
+(TODO: The above does not yet contain the custom values needed for NHCBs. Update
````
Oh nice, so it will be possible to do nhcb on scrape? That's would be nice and somehow along the https://github.com/prometheus/OpenMetrics/issues/283 ideas.

bwplotka replied:
Ok, explained in NHCB section, makes sense!

bwplotka commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2416 @@
+approach, which might be improved in the future to find a better trade-off
+between chunk size and compression ratio.
+
+### Counter reset considerations
```
This is quite complex. What do you think about switching to Created Timestamp feature for it? Would it simplify it? We will have it for all counters soon.

beorn7 replied:
CT is in itself a new and experimental feature. I wouldn't conflate them.

Even if CT were already established, I don't think it can be relied upon for NH ingestion. We need to detect counter reset as part of the encoding anyway. CT is more an additional source of information: We can mark (and store) a counter reset even if it is not detectable by conventional means, see below.

bwplotka commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2416 @@
+The counter reset information of the synthetic zero sample is always set to
+`CounterReset`. (TODO: Currently, Prometheus probably sets it to
+`UnknownCounterReset` for the first sample of a series, which is not wrong, but
+I think setting it to `CounterReset` makes more sense.)
```
Yea, I think we want to switch to native CT handling, not synthetic zeros. So switching counter reset to CT would be one way of doing this, going forward 🤔

beorn7 replied:
See above.

bwplotka reviewed (commented):
Nice! Quite an effort and level of the details. Thanks for this work!

Added a few comments, I wonder if we could prepare the reader for nhcb dimension in the glossary and flavors 🤔 Also should we version this spec (like we do with PRW) 

I also have a general questions to counter reset hints. We could simplify by using the CT feature on native histograms instead of reset hints. Those feature do the same, just CT allows a bit more information to be passed (when exactly reset happened). I understand the reset hints are implemented and "baking" for stability, so perhaps it should be improved in some 2.0 🤔

krajorama reviewed (approved):
LGTM as starting point, awesome work .

beorn7 commented:
Thanks @krajorama & @bwplotka 

@bwplotka : I'll address the individual comments soon. Just a few general things in advance:

> I wonder if we could prepare the reader for nhcb dimension in the glossary and flavors 

I tried to include NHCB wherever it was already possible. NHCB is a schema, though, and thus an already defined dimension of the data model. I wouldn't put it under "flavor" (which is anyway a bit of a makeshift solution to have a word for the counter/gauge and integer/float dichotomies).

WRT glossary: My intention was to nail down the words that the reader might get wrong (because of historical baggage ("sparse histogram") or confusion with previously existing concepts (classic vs. native histograms)). Adding all the terms newly introduced by this document would be quite repetitive, as one of the main purposes of this document is to introduce and explain those terms.

> Also should we version this spec (like we do with PRW)

Maybe. But not while it is not really done. It is kind-of 0.x right now. But even once it is finalized, I'm not sure if versioning makes sense because this is more a concept document than strictly a spec. Let's discuss once we have declared native histograms a stable feature.

> I also have a general questions to counter reset hints. We could simplify by using the CT feature on native histograms instead of reset hints. Those feature do the same, just CT allows a bit more information to be passed (when exactly reset happened). I understand the reset hints are implemented and "baking" for stability, so perhaps it should be improved in some 2.0

Counter resets are more deeply baked into NH. For example, it is impossible to have a counter reset in the middle of a chunk, based on the way the encoding works. Thus, we cannot rely on an external feature, that might or might not be present, to mark counter resets. Or even if we did rely on it, we would still store the counter reset implicitly. In any case, we have to see how CT shakes out in practice, how clock skew is handled and all of that.

beorn7 commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2416 @@
+
+# Native Histograms
+
+Native histograms were introduced as an experimental feature in November 2022.
```
It feels more like 0.1 (or 0.0.1 ;). Once we have addressed all (or at least most of) the TODOs here, and declared NH a stable feature, we might think about versioning. Right now, things are just too much in flux. (There are the git hashes, in any case.)

But I'm not even sure if there should be one version for all of NH. There is the exposition format (with its own versioning), there is RW (with its own versioning), there are chunks (with their own versioning), etc.

bwplotka replied:
Perhaps make it 1.0-rc.0 then, like we iterate over PRW 2.0? https://prometheus.io/docs/specs/remote_write_spec_2_0/ 🤔

beorn7 replied:
I would argue that version numbers aren't the right thing for this pseudo spec. It's way too large of a surface to version. As said above, we should version formats and protocols, not this entire beast.

(I'm also somewhat hesitant to add this to the "specification" section. It could as well be in the "concepts" section. Not saying I want to change it now, just to stress the point that this is only partially a spec.)

beorn7 commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2416 @@
+float samples.
+
+Native histograms have the following key properties:
+1. A sparse bucket representation, allowing (near) zero cost for empty buckets.
```
Yeah, weird. I'll try with the blank line.

beorn7 commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2416 @@
+Native histograms have the following key properties:
+1. A sparse bucket representation, allowing (near) zero cost for empty buckets.
+2. Coverage of the full float64 range of values.
+3. No configuration of bucket boundaries during instrumentation.
```
True. I'll add a clarifying sentence.

beorn7 commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2416 @@
+   all histograms using those schemas.
+6. An efficient data representation for both exposition and storage.
+
+Compared to the previously existing “classic” histograms, native histograms
```
will also add a note for clarification.

beorn7 commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2416 @@
+
+## Glossary
+
+- A __native histogram__ is an instance of the new complex sample type
```
I think a native histogram is a native histogram of any schema. We should rather clarify at other locations (like above) rather than somehow suggesting that an NHCB is not a native histogram. (I tried to do that already, but as you have found, I wasn't entirely consistent.)

bwplotka replied:
I think we are on the same page then, thanks!

beorn7 commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2416 @@
+contains the following components, which are described in detail in dedicated
+sections below:
+
+- A _schema_ to identify the method of determining the boundaries of any given
```
I wouldn't call it a part of the data model. It is, on the one hand, a result of the way we encode it. On the other hand, it is an implementation detail to not do the expensive counter reset detection too often.

bwplotka replied:
fair, thanks!

beorn7 commented:
One more thought about the terms I put into the glossary: I think I understand my reluctance to put all the new terms into it better now. What I want is that the reader can just read through the whole document in order, including the glossary. This means it should only contain terms the reader is already somewhat familiar with. A "conventional" glossary is more something you would put at the end of the document where the reader can refer to if they don't remember a certain term in the middle of the text. But the glossary I'm using here is more along the lines of "let's get some common misunderstandings out of the way first".

fatsheep9146 commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2422 @@
+- The next negative bucket (index _i_+1 relative to the bucket from the
+  previous item) has an upper exclusive limit of `MinFloat64` and an lower
+  inclusive limit of `-Inf`. (It could be called a _negative overflow bucket_.)
+- Buckets beyond the `+Inf` and `-Inf` buckets described above MUST NOT be used.
```
Using the `positive overflow bucket` and `negative overflow bucket` might be more accurate than `+Inf` and `-Inf` buckets?

beorn7 replied:
Those terms are only introduced in parentheses with "they could be called". I'd rather refer to the limits as that's firmly established before. It's also very accurate because the only observed value that will go into the +Inf bucket is, well, +Inf, and likewise for -Inf.

fatsheep9146 commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2422 @@
+In a similar spirit, there are situation where excluding every unpopulated
+bucket from the bucket list is most efficient, but in other situations, it
+might be better to reduce the number of spans by representing small numbers of
+unpopulated buckets explicitly.
```
I don't fully understand here. How could we reduce the number of spans by representing small numbers of unpopulated buckets explicitly? Do you mean combine two successive spans into a new span with some unpopulated buckets in it?

beorn7 commented:
@bwplotka is there anything left to address right now? Can I merge this?

bwplotka reviewed (approved):
> @bwplotka is there anything left to address right now? Can I merge this?

All good, thanks for an amazing and deep work here! 💪🏽

beorn7 commented:
There we go. And this is just the beginning. 43 TODOs to address… :sweat_smile:

fatsheep9146 commented on content/docs/specs/native_histograms.md:
```diff
@@ -0,0 +1,2422 @@
+  from zero, a relatively large threshold of the zero bucket helps to avoid
+  many high resolution buckets for a range that is not of interest.
+
+The threshold of the zero bucket SHOULD coincide with a boundary of a regular
```
Do this should be "SHOULD NOT" instead of "SHOULD"? I take "coincide with" as "same or overlapped".
